experimental/Schemes/Resolution_structure.jl

```diff
@@ -1,54 +1,138 @@
+  # boolean flags
+  resolves_sing::Bool                            # domain not smooth yet?
+  is_trivial::Bool                               # codomain already smooth?
+
+  # fields for caching, to be filled during desingularization
+  ex_div::Vector{AbsIdealSheaf}                  # list of exc. divisors arising from individual steps
+                                                 # in domain(maps[end]) -- at least in the end
+  normal_steps::Vector{Int}                      # index of parts of preimage of singular locus 
+                                                 # not arising from blow-ups
+  ## CAUTION: ex_div[normal_steps[i]] need not even be of pure codim 1, i.e. a divisor, 
+  ##                                  as long as resolves_sing == false
+  ##          lower dimensional components of it belong to future centers of blow ups 
+
+
+  # fields for caching to be filled a posteriori (on demand, only if partial_res==false)
+  underlying_morphism::CompositeCoveredSchemeMorphism{DomainType, CodomainType}
+  exceptional_divisor::WeilDivisor          
+
+  function MixedBlowUpSequence(maps::Vector{<:AbsCoveredSchemeMorphism})
+    n = length(maps)
+    for i in 1:n
+      @assert all(x->((x isa BlowupMorphism) || (x isa NormalizationMorphism)), maps) "only blow-ups and normalizations allowed"
+    end
+    for i in 1:n-1
+      @assert domain(maps[i]) === codomain(maps[i+1]) "not a sequence of morphisms"
+    end
+    resi = new{typeof(domain(maps[end])),typeof(codomain(first(maps)))}(maps)
+    resi.normal_steps = [i for i in 1:n if maps[i] isa NormalizationMorphism]
+    return resi
+  end
+
+end
 
 
 ##################################################################################################
 # getters
 ##################################################################################################
+underlying_morphism(phi::NormalizationMorphism) = CoveredSchemeMorphism(phi.X,phi.Y,phi.f)
+domain(phi::NormalizationMorphism) = phi.X
+codomain(phi::NormalizationMorphism) = phi.Y
 morphisms(phi::AbsDesingMor) = copy(phi.maps)
 morphism(phi::AbsDesingMor,i::Int) = copy(phi.maps[i])
 last_map(phi::AbsDesingMor) = phi.maps[end]
-exceptional_divisor_list(phi::BlowUpSequence) = phi.ex_div  ## derzeit Liste von Eff. Cartier Div.
+
+exceptional_divisor_list(phi::BlowUpSequence) = phi.ex_div
+
+## entries of ex_div corresponding to normalization steps are only exceptional divisors at the very end
+## so only return them at the very end or on specific demand 
+function exceptional_divisor_list(phi::MixedBlowUpSequence, seq_unclean::Bool=false)
+  phi.resolves_sing || seq_unclean || error("only available for completed resolution")
+  return phi.ex_div
+end
+
+normal_steps(phi::MixedBlowUpSequence) = phi.normal_steps
+ 
 embeddings(phi::BlowUpSequence) = phi.embeddings
 
-## do not use!!! (for forwarding and certain emergenies)
+## do not use!!! (for forwarding and certain emergencies)
```

Comment:
I wonder why one should not use this?

Reply:
This is a leftover from before Matthias brushed it up a bit. I will delete it.

Reply:
done